### PR TITLE
Images fail to load periodically on ikea.com

### PR DESCRIPTION
**Pull request by stwrt**
#### 5e60d4e7a72af29c36f4874cd993016829422b23
<pre>
Images fail to load periodically on ikea.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=248048">https://bugs.webkit.org/show_bug.cgi?id=248048</a>
rdar://101657305

Reviewed by Chris Dumez.

Adding a quirk to address images showing up as grey boxes on ikea.com.
Disabling lazy image loading addresses an issue that was found
in gatsby. This issue has been addressed in a future version of gatsby.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::lazyImageLoadingEnabled const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldExposeShowModalDialog const):
(WebCore::Quirks::shouldDisableLazyImageLoading const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/256854@main">https://commits.webkit.org/256854@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3272fbba99c28544a907f389cdf3029d7ec508e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96995 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106524 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6491 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34997 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89387 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103218 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102666 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83610 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31894 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74785 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/297 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/279 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4741 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5071 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1513 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40801 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/dom/Document.cpp [hunk 1]
```diff
@@ -9334,6 +9334,12 @@ void Document::sendReportToEndpoints(const URL& baseURL, const Vector<String>& e
     }
 }
 
+bool Document::lazyImageLoadingEnabled() const
+{
+    auto result = m_settings->lazyImageLoadingEnabled() && !m_quirks.get().shouldDisableLazyImageLoading();
```
Comment (cdumez) [inline Source/WebCore/dom/Document.cpp]: `m_quirks->`

Comment (cdumez) [reply Source/WebCore/dom/Document.cpp]: No need for a local `result` variable.

File: Source/WebCore/page/Quirks.cpp [hunk 2]
```diff
@@ -1568,11 +1569,36 @@ bool Quirks::shouldExposeShowModalDialog() const
 {
     if (!needsQuirks())
         return false;
+
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.cpp]: Unnecessary change.

File: Source/WebCore/page/Quirks.cpp [hunk 3]
```diff
@@ -1568,11 +1569,36 @@ bool Quirks::shouldExposeShowModalDialog() const
...
+    // Quirk only applies for IKEA.com.
+    // Searching for <meta name="generator" content="Gatsby 4.24.1" />
+    // Images will have a grey box over them when loaded lazily in large percentage of page loads.
+    // This issue is not observed when lazy image loading is disabled, and has been fixed in future Gatsby versions.
+    auto* metaElement = m_document->getElementsByTagName("meta"_s).get().namedItem("generator"_s);
+    
+    if (RegistrableDomain(m_document->url()).string() == "ikea.com"_s 
+        && metaElement 
+        && metaElement->getAttribute("content"_s) == "Gatsby 4.24.1"_s)
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.cpp]: Per WebKit coding style, if a condition is on multiple lines, you need to use curly brackets.

File: Source/WebCore/page/Quirks.cpp [hunk 4]
```diff
@@ -1568,11 +1569,36 @@ bool Quirks::shouldExposeShowModalDialog() const
...
+    if (m_shouldDisableLazyImageLoading)
+        return m_shouldDisableLazyImageLoading.value();
+
+    m_shouldDisableLazyImageLoading = false;
+
+    // Quirk only applies for IKEA.com.
+    // Searching for <meta name="generator" content="Gatsby 4.24.1" />
+    // Images will have a grey box over them when loaded lazily in large percentage of page loads.
+    // This issue is not observed when lazy image loading is disabled, and has been fixed in future Gatsby versions.
+    auto* metaElement = m_document->getElementsByTagName("meta"_s).get().namedItem("generator"_s);
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.cpp]: Use `->` instead of `.get()`

File: Source/WebCore/page/Quirks.h [hunk 5]
```diff
@@ -165,6 +166,8 @@ class Quirks {
     bool shouldEnableApplicationCacheQuirk() const;
     bool shouldEnableFontLoadingAPIQuirk() const;
     bool needsVideoShouldMaintainAspectRatioQuirk() const;
+
+    bool shouldDisableLazyImageLoading() const;
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.h]: Other functions seem to have "Quirk" in their name, let's follow the pattern.

**Review by cdumez — COMMENTED** [review]
You mentioned on Slack that this patch wasn't working yet so maybe this PR should be marked as Draft?

**Comment by cdumez** [comment]
> Images fail to load periodically on ikea.com
> https://bugs.webkit.org/show_bug.cgi?id=248048

Please put radar number here too.

**Comment by stwrt** [comment]
> > Images fail to load periodically on ikea.com
> > https://bugs.webkit.org/show_bug.cgi?id=248048
> 
> Please put radar number here too.

👍

File: Source/WebCore/dom/Document.cpp [hunk 6]
```diff
@@ -9334,6 +9334,11 @@ void Document::sendReportToEndpoints(const URL& baseURL, const Vector<String>& e
     }
 }
 
+bool Document::lazyImageLoadingEnabled() const
+{
+    return (m_settings->lazyImageLoadingEnabled() && !m_quirks->shouldDisableLazyImageLoadingQuirk());
```
Comment (cdumez) [inline Source/WebCore/dom/Document.cpp]: No need for the surrounding parentheses.

File: Source/WebCore/page/Quirks.cpp [hunk 7]
```diff
@@ -1501,7 +1502,6 @@ bool Quirks::needsVideoShouldMaintainAspectRatioQuirk() const
 {
     if (!needsQuirks())
         return false;
-
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.cpp]: Unnecessary change, please revert.

File: Source/WebCore/page/Quirks.cpp (hunk 2 above)
Comment (cdumez) [reply Source/WebCore/page/Quirks.cpp]: Unnecessary change. This wasn't resolved.

File: Source/WebCore/page/Quirks.cpp [hunk 8]
```diff
@@ -1515,11 +1515,36 @@ bool Quirks::shouldExposeShowModalDialog() const
...
+
+    m_shouldDisableLazyImageLoadingQuirk = false;
+
+    // Images will have a grey box over them when loaded lazily in large percentage of page loads.
+    // This issue is not observed when lazy image loading is disabled, and has been fixed in future Gatsby versions.
+    // This quirk only needs to be applied to IKEA.com when "<meta name="generator" content="Gatsby 4.24.1" />" is present.
+
+    auto* metaElement = m_document->getElementsByTagName("meta"_s)->namedItem("generator"_s);
+    
+    if ((RegistrableDomain(m_document->url()).string() == "ikea.com"_s)
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.cpp]: This check needs to happen earlier. Why are we even traversing the DOM tree to find meta tags if we're not on IKEA?

File: Source/WebCore/page/Quirks.cpp [hunk 9]
```diff
@@ -1515,11 +1515,36 @@ bool Quirks::shouldExposeShowModalDialog() const
...
+        && (metaElement)
+        && (metaElement->getAttribute("content"_s) == "Gatsby 4.24.1"_s))
```
Comment (cdumez) [inline Source/WebCore/page/Quirks.cpp]: As I mentioned in my earlier review, missing the curly brackets.

File: Source/WebCore/page/Quirks.cpp (hunk 7 above)
Comment (brentfulgham) [reply Source/WebCore/page/Quirks.cpp]: @stwrt Whitespace changes should be omitted from the change.

**Review by cdumez — CHANGES_REQUESTED** [review]
There is evidence this is a recent WebKit regression. Quirks are to work around a site bug, not a WebKit bug. So why are we doing a quirk?

Also, it seems lazy image loading got enabled before the regression point?

File: Source/WebCore/page/Quirks.cpp (hunk 8 above)
Comment (stwrt) [reply Source/WebCore/page/Quirks.cpp]: Good point fixed.

**Review by cdumez — APPROVED** [review]
LGTM since Brandon tells me this is not a WebKit regression (The radar probably needs updating to clarify that).

File: Source/WebCore/page/Quirks.cpp [hunk 10]
```diff
@@ -1522,4 +1523,29 @@ bool Quirks::shouldExposeShowModalDialog() const
     return *m_shouldExposeShowModalDialog;
 }
 
+bool Quirks::shouldDisableLazyImageLoadingQuirk() const
+{
```
Comment (karlcow) [inline Source/WebCore/page/Quirks.cpp:1527]: In a local branch of mine, I have started to add comments for the quirks, because after a while it becomes very hard to understand where it comes from.

It probably be a good first step to refactor latter with messages in the console. 

My format so far has been:

```suggestion

// ikea.com https://bugs.webkit.org/show_bug.cgi?id=248048
bool Quirks::shouldDisableLazyImageLoadingQuirk() const
{
```

Comment (stwrt) [reply Source/WebCore/page/Quirks.cpp:1527]: I was planning to add a link to the current Ikea discussion.

For the quirks I think I would like a bit more information then just a link to a bugzilla. Having information on the background of why this quirk exists, and information on how to eventually remove it would be nice too. @karlcow thoughts?

Comment (karlcow) [reply Source/WebCore/page/Quirks.cpp:1527]: More would be better. We could come up with a kind of format where we could normalize step by step the other quirks too, but just start with this one. 
Something that can be easily parsed.
